Move from State abstraction to direct zk access with scala based Futures

aquamatthias opened the pull request:
@kolloch Please review.

Discussion points:
-package of Store abstraction
-Future timeouts 

Fixes #1539

kolloch commented on src/main/proto/marathon.proto:
```diff
@@ -174,3 +174,12 @@ message TaskFailure {
+
+// Describes a state entry, a versioned (via a UUID) key/value pair.
+// Copied from libmesos (src/messages/state.pro) to ensure
+// compatibility with ZooKeeperState from libmesos.
```
Nice comment! in mesos.proto they use `/**... */` though.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,127 @@
+}
+
+object ZKStore {
+  implicit class Twitter2Scala[T](val twitterF: TWFuture[T]) extends AnyVal {
```
I am somewhat an implicit conversion hater.

If I want to understand .asScala above, I need to check out the code and use my IDE. Otherwise I'll have a hard time finding the relevant code.

aquamatthias replied:
Understand the point. I think/hope that explicit implicit makes the code more readable.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,127 @@
+    root.getChildren().map(_.children.map(_.name)).asScala
+  }
+
+  private def mapFutureResult[A, B](value: TWFuture[A], errorMessage: String)(fn: A => TWFuture[B]): Future[B] = {
```
maybe make errorMessage a function? It should only be evaluated in the error case.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,127 @@
+    val zk = zkEntity(entity)
+    def update(nodeData: ZNode.Data): Future[ZKEntity] = {
+      val future: TWFuture[Data] = zk.node.setData(zk.data.toProto.toByteArray, nodeData.stat.getVersion)
+      mapFutureResult(future, s"Can not update entity $entity") { data =>
```
I would map to a scala future as early as possible an then use the well known methods on the scala future.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,127 @@
+  override def delete(key: ID): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    mapFutureResult(node.exists(), s"Can not delete entity $key") { data =>
```
why do an exists first? Can't we just catch errors of the delete? Saves round-trips for us.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,127 @@
+    * Fetch data and return entity.
+    * The entity is returned also if it is not found in zk, since it is needed for the store operation.
+    */
+  override def load(key: ID): Future[ZKEntity] = {
```
Does it cause a lot of changes if we change that to Future[Option[ZKEntity]]?

kolloch commented on src/main/scala/mesosphere/util/state/memory/InMemoryStore.scala:
```diff
@@ -0,0 +1,45 @@
+  * This is intended only for tests - do not use in production!
+  * @param ec the execution context to use.
+  */
+class InMemoryStore(implicit val ec: ExecutionContext = ThreadPoolContext.context) extends PersistentStore {
```
I would be nice if InMemory store also checked the version on updates.

aquamatthias replied:
Added that ability.

kolloch commented on src/main/scala/mesosphere/marathon/state/PersistenceStore.scala:
```diff
@@ -4,16 +4,17 @@ import scala.concurrent.Future
 
 trait PersistenceStore[T] {
 
+  type Read = () => T //returns deserialized T value.
```
I think I would understand the intention better if this was called `Deserialize`

aquamatthias commented:
@kolloch 

-ZKStore: moved from TwitterFuture to ScalaFuture early. Subsequent handling relies on standard types.
-PersistentStore.load will return an Option[Entity] and not an Entity.
 that implies, that create and update has to be handled explicitly
 I added the functionality and rewrote the code, that relied on the implicit entity creation.
-EntityStore.{store, modify} will return no Option, but either succeed or fail
 Rewrote the functionality, that handled the Option case (which generally was considerer an error)
-MarathonStore: clean up of the existing functionality

There is the ability to switch the backend store via the hidden flag --internal_store_backend.
I used this for testing backward compatibility - not sure, if we should remove it.

I removed the zookeeper authentication flags/authentication.
The PR is already too big. Can add this functionality, once this is landed.

kolloch commented on src/main/scala/mesosphere/util/state/mesos/MesosStateStore.scala:
```diff
@@ -0,0 +1,49 @@
+
+  override def load(key: ID): Future[Option[PersistentEntity]] = {
+    futureToFuture(state.fetch(key)).map { variable =>
+      if (variable.value().length > 0) Some(MesosStateEntity(key, variable)) else None
```
Maybe `.nonEmpty` is better? (Also see below)

Maybe this also deserves an explanation.

aquamatthias replied:
I thought you were an implicit hater ... :)
I replace that with a small helper function.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+
+  private def exceptionTransform[T](errorMessage: String): PartialFunction[Throwable, T] = {
+    case ex: KeeperException => throw new StoreCommandFailedException(errorMessage, ex)
+    case ex: Throwable       => throw ex
```
`NonFatal(ex)` better?

aquamatthias replied:
Good catch - it is better.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    node.getData().asScala
+      .map{ data => Some(ZKEntity(node, ZKData(data.bytes), Some(data.stat.getVersion))) }
```
Being picky. Space before `{` as below?

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+    */
+  override def load(key: ID): Future[Option[ZKEntity]] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
```
have the node path in the error message

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+
+  override def create(key: ID, content: Array[Byte]): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
```
node path in the error message

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+    }
+  }
+
+  private def createPath(path: ZNode): ZNode = {
```
This uses blocking which is fine (in my opinion) in initialization but shouldn't be used elsewhere.

Thus I would name this method `createPathBlocking` or something like this.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+  private def zkEntity(entity: PersistentEntity): ZKEntity = {
+    entity match {
+      case zk: ZKEntity => zk
+      case _            => throw new IllegalArgumentException("Can not handle this kind of entity")
```
Add type of the argument to the error message.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+        .flatMap { _.getData().map{ data => zk.copy(version = Some(data.stat.getVersion)) }.asScala }
+        .recover(exceptionTransform(s"Can not update entity $entity"))
+    }
+    zk.version.map(update).getOrElse(create())
```
The decision whether to save(= update) or to create (= insert) is now actually handled in `MarathonStore.modify`. Thus this code can go. Or did I misunderstand something?

aquamatthias replied:
Removed the create version.

kolloch commented on src/main/scala/mesosphere/util/state/zk/ZKStore.scala:
```diff
@@ -0,0 +1,149 @@
+    * The entity will be either created or updated, depending on the read state.
+    * @return Some value, if the store operation is successful otherwise None
+    */
+  override def save(entity: PersistentEntity): Future[ZKEntity] = {
```
Maybe rename to `update`, see comment below.

aquamatthias replied:
Agree, it should be CRUD: Create/Update/Delete

kolloch commented:
Review still in progress. Looks nice so far! Notes for myself:

* Looked at ZKStore
* Looked at InMemoryStore
* Ran integration tests with mesos_zk, they failed, investigate.

What does native in `NativeConnector` mean?

kolloch commented on src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala:
```diff
@@ -153,7 +152,7 @@ class MarathonSchedulerActor(
       withLockFor(appId) {
         val promise = Promise[Unit]()
-        val tasksToKill = taskIds.flatMap(taskTracker.fetchTask(appId, _)).toSet
+        val tasksToKill = taskIds.flatMap(taskTracker.fetchTask(appId, _))
```
Just a note to myself that this is OK ;)

kolloch commented on src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala:
```diff
@@ -416,7 +415,7 @@ class SchedulerActions(
   def stopApp(driver: SchedulerDriver, app: AppDefinition): Future[_] = {
     appRepository.expunge(app.id).map { successes =>
       if (!successes.forall(_ == true)) {
```
It is not your code but WTF?? they are always all true but errors are not handled correctly

(Not part of your PR)...

kolloch commented on src/main/scala/mesosphere/marathon/ZookeeperConf.scala:
```diff
@@ -12,23 +13,19 @@ trait ZookeeperConf extends ScallopConf {
-    default = Some("localhost:2181"))
-
-  @Deprecated
-  lazy val zooKeeperPath = opt[String]("zk_state",
```
You should remove this from the Docs as well (see rest-api.md and developing-vm.md)

kolloch replied:
And add a comment in changelog.md just in case.

aquamatthias replied:
Done. Thanks for the hint.

kolloch commented on src/main/scala/mesosphere/marathon/event/http/HttpEventModule.scala:
```diff
@@ -49,7 +50,7 @@ class HttpEventModule(httpEventConfiguration: HttpEventConfiguration) extends Ab
   def provideSubscribersKeeperActor(conf: HttpEventConfiguration,
                                     system: ActorSystem,
-                                    store: MarathonStore[EventSubscribers]): ActorRef = {
+                                    store: EntityStore[EventSubscribers]): ActorRef = {
```
I like it :)

kolloch commented on src/main/scala/mesosphere/marathon/event/http/SubscribersKeeperActor.scala:
```diff
@@ -2,43 +2,38 @@ package mesosphere.marathon.event.http
 import scala.concurrent.Future
 
-class SubscribersKeeperActor(val store: MarathonStore[EventSubscribers]) extends Actor with ActorLogging {
+class SubscribersKeeperActor(val store: EntityStore[EventSubscribers]) extends Actor with ActorLogging {
```
Your changes made me think that MarathonStore can actually be package private. What do you think?

aquamatthias replied:
The guice binder (MarathonModule) needs access to this class.
It would need more actions to accomplish this, so I will leave that in current state.

kolloch commented on src/main/scala/mesosphere/marathon/state/EntityStore.scala:
```diff
@@ -0,0 +1,20 @@
+  type Deserialize = () => T //returns deserialized T value.
+  type Update = Deserialize => T //Update function Gets an Read and returns the (modified) T
+
+  def fetch(key: String): Future[Option[T]]
```
Some docs?

aquamatthias replied:
Added class docs.

kolloch commented on src/main/scala/mesosphere/marathon/state/MarathonStore.scala:
```diff
@@ -1,139 +1,75 @@
-    implicit val timeout: BackToTheFuture.Timeout = BackToTheFuture.Timeout(Duration(conf.marathonStoreTimeout(),
-      MILLISECONDS)))
-    extends PersistenceStore[S] {
+    val store: PersistentStore,
```
Thank you for not extending PersistenceStore!!!

kolloch replied:
But why is store a val?

aquamatthias replied:
Mmmh - I removed the val and it still compiles, so I don't think it is needed.

kolloch commented on src/main/scala/mesosphere/marathon/state/MarathonStore.scala:
```diff
@@ -1,139 +1,75 @@
-      case None =>
-        throw new StorageException(s"Failed to read $key, does not exist, should have been created automatically.")
+    store.load(prefix + key).map {
+      _.map{ entity =>
```
This is so much better than before!

format nitpicking, space after `map`?

aquamatthias replied:
Changed it. Is there a way to force this via scalariform?

kolloch replied:
I don't know. Do you? :)

kolloch commented on src/main/scala/mesosphere/marathon/state/MarathonStore.scala:
```diff
@@ -1,139 +1,75 @@
+      _.map{ entity =>
+        bytesRead.update(entity.bytes.length)
+        stateFromBytes(entity.bytes)
+      }
```
A recover that adds context and rethrows an exception would be nice.

For example, if serialization fails, it would be very nice to know for what kind of entity and for which ID.

kolloch commented on src/main/scala/mesosphere/marathon/state/MarathonStore.scala:
```diff
@@ -1,139 +1,75 @@
     newState().mergeFromProto(bytes)
   }
+
+  private def withLock[T](key: String)(future: => Future[T]): Future[T] = {
```
Really nice that you have extracted that.

Optimally, LockManager would return a LockManger object containing this method, though.

aquamatthias replied:
Like the idea. Still think, there is certain functionality missing in the standard Future companion, perhaps we can collect such functionality into such one.

kolloch commented on src/main/scala/mesosphere/util/state/mesos/MesosStateStore.scala:
```diff
@@ -0,0 +1,49 @@
+    }
+  }
+
+  override def allIds(): Future[Seq[ID]] = futureToFuture(state.names()).map(_.asScala.toSeq)
```
Doesn't that through an NPE if the Zookeeper is empty currently? See the code you removed from `MarathonStore.names()` which is used in the Migration code. Or am I missing something?

aquamatthias replied:
Thanks for remembering - that one is really ugly...

kolloch commented on src/main/scala/mesosphere/marathon/state/MarathonStore.scala:
```diff
@@ -1,139 +1,75 @@
-
-    res
+  def expunge(key: String): Future[Boolean] = withLock(key) {
+    store.delete(prefix + key) map { _ => true }
```
I know it is ugly and I do not really want to burden you with correcting all this in this PR but to keep the old semantics: 

Don't you not at least need a `recover { case NonFatal(e) => log something; false }`?

But better actually distinguish between errors and missing nodes here.

kolloch replied:
I added a note to `PersistentStore.delete` to return a `Future[Option[...]]` so you can map it to true false "correctly".

aquamatthias replied:
I don't like the option return type here. The future already has the notion of success/failure, which is somehow translated to true/false or would be some/none in this case. Either Boolean as also Option say: you have to handle the success of the operation yourself. I will make it Future[Unit], which signals: delete was ok or not. What do you think?

kolloch replied:
I don't agree.

You typically want to handle the case "the node does not exist" differently from "I couldn't connect to zookeeper".

A compromise would be to define an Exception which is only thrown in the "the node does not exist case".

aquamatthias replied:
The question is, do we need an exists method somewhere. We could put that on the PersistentEntity if needed. The notion of delete(entity) = Success(None) reminds me on Some(null)...

aquamatthias replied:
Since there is no type to wrap into an option, I will take Future[Boolean]
Success: The entity was deleted(true) or is not existent(false) => in any case it is gone
Failure: The entity could not get deleted

kolloch replied:
Sounds good. The only thing that is missing then is the ability to take a look at the old instance (since we retrieve it anyway). But since we do not need that...

Besides: `Success(None)` is nothing like `Some(null)` but may make a lot of sense. `Future[Option[...]]` is common and useful.

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,71 @@
+    * @return either the entity or a failure.
+    *         In case of a storage specific failure, a StoreCommandFailedException is thrown.
+    */
+  def delete(key: ID): Future[PersistentEntity]
```
I guess this should return a `Future[Option[PersistentEntity]]` to distinguish deletes from noops (because the entity is already gone).

Also see the caller `expunge` which would benefit from that.

aquamatthias replied:
See my comment on expunge.

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,71 @@
+  /**
+    * Expunge the entity with given id.
+    * @param key the key identifier
+    * @return either the entity or a failure.
```
the entity prior to the delete

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,71 @@
+  def load(key: ID): Future[Option[PersistentEntity]]
+
+  /**
+    * Create a new entity with given id and content.
```
Document: Results in error if the entity already exists.

aquamatthias replied:
Added a note on that.

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,71 @@
+    * @return either the entity or a failure.
+    *         In case of a storage specific failure, a StoreCommandFailedException is thrown.
+    */
+  def save(entity: PersistentEntity): Future[PersistentEntity]
```
Document: Optimistic locking. Save will fail, if the given entity doesn't correspond to current version.

kolloch replied:
Maybe call `update` instead of `save` to make it clear that we deal with existing entities?

kolloch commented on src/main/scala/mesosphere/util/state/FrameworkIdUtil.scala:
```diff
@@ -0,0 +1,58 @@
+  * @param key The key to store the framework ID under
+  */
+
+class FrameworkIdUtil(val state: PersistentStore, val key: String = "frameworkId") {
```
Somehow I do not find the deleted FrameworkIdUtil in the changes, strange. 

Why do you use `val` here? And wouldn't it be to use `MarathonStore` instead of `PersistentStore`?

aquamatthias replied:
This class lives in mesos-utils and depend on the State abstraction.
I basically copied this class and made the needed adaptions.
Yes: it should also use MarathonStore - ok I will change this also.

kolloch replied:
Ah, ok, thanks!

I know this PR gets large.

kolloch commented on src/main/scala/mesosphere/marathon/MarathonConf.scala:
```diff
@@ -159,4 +159,12 @@ trait MarathonConf extends ScallopConf with ZookeeperConf with IterativeOfferMat
+  lazy val internalStoreBackend = opt[String]("internal_store_backend",
+    descr = "The backend storage system to use. One of zk, mesos_zk",
+    hidden = true,
+    validate = Set("zk", "mesos_zk").contains,
```
It would be occasionally useful to also allow `in_memory`. For example, to test whether persistence is our bottleneck.

kolloch commented on src/main/scala/mesosphere/util/state/memory/InMemoryStore.scala:
```diff
@@ -0,0 +1,57 @@
+  */
+class InMemoryStore(implicit val ec: ExecutionContext = ThreadPoolContext.context) extends PersistentStore {
+
+  var entities = Map.empty[ID, InMemoryEntity]
```
It would be nice to allow concurrency (for testing).

aquamatthias replied:
Now use concurrent.TrieMap.

kolloch replied:
:+1:

kolloch commented on src/test/scala/mesosphere/util/state/zk/ZKStoreTest.scala:
```diff
@@ -0,0 +1,124 @@
+
+import scala.collection.JavaConverters._
+
+class ZKStoreTest extends IntegrationFunSuite with BeforeAndAfterAll with BeforeAndAfter with Matchers {
```
Nice Test!

It would be great if you created an `AbstractStoreTest` or something similar and then instantiated with your three different store implementations.

kolloch replied:
Of course, some of the tests are backend specific but many are general.

kolloch commented:
Now, I have looked at everything at least once. It's really an improvement to what we have right now and I can't wait to merge the final PR.

Here some comments in addition the inline comments.

It would be cool, if we could run the integration tests against all three possible storage backends on teamcity.

It would also be nice to test the compatibility of the storage formats for both Zookeeper backends.

Does the Mesos Zookeeper state abstraction contain compatibility code for older formats? Then we would need to support older formats, too, or document that we do not.

aquamatthias commented:
Incorporated the suggested changes - they were very helpful, Thanks!

1) I investigated some time for Future Timeouts - it probably make sense to add that before release
2) I ran the integration tests against the 3 possible storage backends. All were green. I am not 100% sure, if we should add that to TeamCity. The build time would get trippled and I hope we use only one option in the future.
3) Some scale testing is still missing - will start to investigate .
4) What do you think to the package question?

kolloch commented:
Cool, nearly there.

Besides my comments, I would really like to see some compatibility testing (if the storage format is compatible).

aquamatthias commented:
Cool.
Just added tests for reading,writing to/from zkStore <-> mesosState. See ZkStoreTest.

aquamatthias commented:
Rebased and squashed.

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,80 @@
+  /**
+    * Mutate this entity with given bytes.
+    */
+  def mutate(bytes: Array[Byte]): PersistentEntity
```
different name and description: This method does not actually modify the instance itself but returns a new instance.

aquamatthias replied:
Renamed the method to withNewContent.

kolloch commented on src/main/scala/mesosphere/util/state/PersistentStore.scala:
```diff
@@ -0,0 +1,80 @@
+  /**
+    * Get the content bytes of this entity.
+    */
+  def bytes: Array[Byte]
```
Maybe use `IndexedSeq[Byte]` instead of `Array[Byte]` to avoid exporting mutable methods.

http://stackoverflow.com/questions/3350421/what-is-most-efficient-way-to-do-immutable-byte-arrays-in-scala

aquamatthias replied:
Scala uses a WrappedArray by default, so that should be not to expensive.

kolloch commented on src/main/scala/mesosphere/util/state/memory/InMemoryStore.scala:
```diff
@@ -0,0 +1,53 @@
+  override def allIds(): Future[Seq[ID]] = Future.successful(entities.keySet.toSeq)
+}
+
+case class InMemoryEntity(id: String, version: Int)(val bytes: Array[Byte] = Array.empty) extends PersistentEntity {
```
This syntax confused me at least: I didn't know that equality/hash is restricted to the first parameter class
of a case class.

aquamatthias replied:
I reverted the change to a simple case class again.

kolloch commented on src/main/scala/mesosphere/util/state/memory/InMemoryStore.scala:
```diff
@@ -0,0 +1,53 @@
+
+case class InMemoryEntity(id: String, version: Int)(val bytes: Array[Byte] = Array.empty) extends PersistentEntity {
+  override def mutate(bytes: Array[Byte]): PersistentEntity = copy()(bytes)
+  def incrementVersion: InMemoryEntity = copy(version = version + 1)(bytes)
```
I like `withIncrementedVersion` better.

aquamatthias commented:
Incorporated your suggestions.

kolloch commented:
I restarted the test but the PR seems to need a rebase anyways.

aquamatthias commented:
Rebased.

aquamatthias commented:
Rebased and squashed.

kolloch commented on changelog.md:
```diff
@@ -36,6 +36,16 @@ The new Marathon now provides an
 where you receive all events conveniently as
 [Server Sent Events](http://www.w3schools.com/html/html5_serversentevents.asp).
 
+#### Abstraction for Persistent Storage added with Zookeeper access directly in the JVM 
```
Capitalization not consistent with other head lines.

aquamatthias replied:
Persistent Storage -> persistent storage

kolloch commented on changelog.md:
```diff
@@ -106,12 +116,17 @@ parameters.
 
-#### Do not delay task launches anymore as a result of failed health checks
-
+####
```
Line break by accident?

aquamatthias replied:
Yes. reverted.

kolloch commented on changelog.md:
```diff
@@ -106,12 +116,17 @@ parameters.
+The command line arguments `zk_hosts` and `zk_state` were deprecated for some time and got removed in this version.
+Use the `--zk` command line argument to define the zookeeper connection string. 
+
 Since health checks typically (depending on configuration) take a while to determine that a task is unhealthy,
```
You interrupted the section with your new section.

aquamatthias replied:
Moved after the section.

kolloch commented on src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala:
```diff
@@ -228,6 +227,16 @@ class MarathonSchedulerService @Inject() (
     driver = None
   }
 
+  def isLeader: Boolean = leader.get()
```
Nope. you readded this by accident, I guess.

kolloch commented on src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala:
```diff
@@ -345,7 +354,7 @@ class MarathonSchedulerService @Inject() (
     timer.schedule(
       new TimerTask {
         def run() {
-          if (leader.get()) {
```
nope, keep `leader.get()`

kolloch commented on src/main/scala/mesosphere/util/LockManager.scala:
```diff
@@ -3,8 +3,38 @@ package mesosphere.util
+    * @tparam T the resulting type of the future.
+    * @return A new future, that gets executed, once the lock is available.
+    */
+  def executeSequential[T](key: String)(future: => Future[T])(implicit ec: ExecutionContext): Future[T]
```
sequentially?

kolloch commented:
Some merge errors, otherwise fine.

kolloch commented:
:+1: !!! LGTM !!!

Waiting for tests.
